File: sdk/dotnet/src/Starfederation.Datastar/HttpHandlers.cs
using System.Text;
using System.Text.Json;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Primitives;
using Microsoft.Net.Http.Headers;

namespace Starfederation.Datastar
{
    /// <summary>
    /// Implementation of ISendServerEvent, for sending SSEs to the HttpResponse
    /// </summary>
    public class ServerSentEventHttpHandlers : ISendServerEvent
    {
        private readonly HttpResponse _httpResponse;
        private readonly List<(string, string)> _additionalHeaders = new();
        private Task? _respondTask = null;

        /// <summary>
        /// Initializes a new instance of the ServerSentEventHttpHandlers class.
        /// </summary>
        /// <param name="httpResponse">The HTTP response.</param>
        public ServerSentEventHttpHandlers(HttpResponse httpResponse)
        {
            _httpResponse = httpResponse;
        }

        /// <summary>
        /// Gets the HTTP response.
        /// </summary>
        public HttpResponse HttpResponse => _httpResponse;

        /// <summary>
        /// Adds a header to the response.
        /// </summary>
        /// <param name="name">The header name.</param>
        /// <param name="content">The header content.</param>
        public void AddHeader(string name, string content)
        {
            _additionalHeaders.Add((name, content));
        }

        /// <summary>
        /// Starts sending the response.
        /// </summary>
        /// <returns>A task that represents the asynchronous start operation.</returns>
        public Task StartResponse()
        {
            _respondTask = Task.Run(async () =>
            {
                SetHeader(_httpResponse, (HeaderNames.ContentType, "text/event-stream"));
                
                if (_httpResponse.HttpContext.Request.Protocol == "HTTP/1.1")
                {
                    //TODO test if this is needed
                    SetHeader(_httpResponse, (HeaderNames.CacheControl, "no-cache"));
                    SetHeader(_httpResponse, (HeaderNames.Connection, "keep-alive"));
                    //SetHeader(_httpResponse, (HeaderNames.KeepAlive, "timeout=60"));
                }
                
                foreach (var header in _additionalHeaders)
                {
                    SetHeader(_httpResponse, header);
                }
                
                // Write the initial response headers
                await _httpResponse.Body.FlushAsync();
                return;

                void SetHeader(HttpResponse httpResponse, (string name, string content) header)
                {
                    if (!httpResponse.Headers.ContainsKey(header.name))
                    {
                        httpResponse.Headers[header.name] = new StringValues(header.content);
                    }
                }
            });
            
            return _respondTask;
        }

        /// <summary>
        /// Sends a server-sent event.
        /// </summary>
        /// <param name="serverSentEvent">The server-sent event to send.</param>
        /// <returns>A task that represents the asynchronous send operation.</returns>
        public async Task SendServerEventAsync(ServerSentEvent serverSentEvent)
        {
            if (_respondTask == null)
            {
                throw new System.InvalidOperationException("StartResponse() must be called before sending events.");
            }
            
            await _respondTask;
            
            var serializedEvent = serverSentEvent.Serialize();
            var bytes = Encoding.UTF8.GetBytes(serializedEvent);
            await _httpResponse.Body.WriteAsync(bytes);
        }
    }

    /// <summary>
    /// Implementation of IReadSignals, for reading the Signals from the HttpRequest
    /// </summary>
    public class SignalsHttpHandlers : IReadSignals
    {
        private readonly HttpRequest _httpRequest;

        /// <summary>
        /// Initializes a new instance of the SignalsHttpHandlers class.
        /// </summary>
        /// <param name="httpRequest">The HTTP request.</param>
        public SignalsHttpHandlers(HttpRequest httpRequest)
        {
            _httpRequest = httpRequest;
            _httpRequest.EnableBuffering();
        }

        /// <summary>
        /// Gets the HTTP request.
        /// </summary>
        public HttpRequest HttpRequest => _httpRequest;

        private async Task<Optional<string>> ReadRawSignalsAsync()
        {
            if (_httpRequest.Method == "GET")
            {
                // Handle GET request (if implemented)
                return Optional<string>.None;
            }

            // Read the request body
            _httpRequest.Body.Position = 0;
                
            using var reader = new StreamReader(_httpRequest.Body, Encoding.UTF8, leaveOpen: true);
            var body = await reader.ReadToEndAsync();
                
            if (string.IsNullOrWhiteSpace(body))
            {
                return Optional<string>.None;
            }
                
            return Optional<string>.Some(body);
        }

        /// <summary>
        /// Reads the signals from the request.
        /// </summary>
        /// <returns>A task that represents the asynchronous read operation. The result contains the signals if present.</returns>
        public async Task<Optional<Signals>> ReadSignalsAsync()
        {
            var rawSignals = await ReadRawSignalsAsync();
            return rawSignals.HasValue 
                ? Optional<Signals>.Some(new Signals(rawSignals.Value)) 
                : Optional<Signals>.None;
        }

        /// <summary>
        /// Reads the signals from the request and deserializes them to the specified type.
        /// </summary>
        /// <typeparam name="T">The type to deserialize to.</typeparam>
        /// <returns>A task that represents the asynchronous read operation. The result contains the deserialized data if present.</returns>
        public Task<Optional<T>> ReadSignalsAsync<T>()
        {
            return ReadSignalsAsync<T>(JsonSerializerOptions.Default);
        }

        /// <summary>
        /// Reads the signals from the request and deserializes them to the specified type using the provided options.
        /// </summary>
        /// <typeparam name="T">The type to deserialize to.</typeparam>
        /// <param name="jsonSerializerOptions">The JSON serializer options to use.</param>
        /// <returns>A task that represents the asynchronous read operation. The result contains the deserialized data if present.</returns>
        public async Task<Optional<T>> ReadSignalsAsync<T>(JsonSerializerOptions jsonSerializerOptions)
        {
            var signals = await ReadRawSignalsAsync();
            
            if (!signals.HasValue)
            {
                return Optional<T>.None;
            }
            
            try
            {
                var deserialized = JsonSerializer.Deserialize<T>(signals.Value, jsonSerializerOptions);
                return deserialized == null 
                    ? Optional<T>.None 
                    : Optional<T>.Some(deserialized);
            }
            catch
            {
                return Optional<T>.None;
            }
        }
    }
}

File: sdk/dotnet/fsharp/src/HttpHandlers.fs
namespace StarFederation.Datastar.FSharp

open System.IO
open System.Text
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.Primitives
open Microsoft.Net.Http.Headers

module JsonSerializerOptions =
    /// JsonSerializerOptions but case insensitive
    let SignalsDefault =
        let options = JsonSerializerOptions()
        options.PropertyNameCaseInsensitive <- true
        options

/// Implementation of ISendServerEvent, for sending SSEs to the HttpResponse
[<Sealed>]
type ServerSentEventHttpHandler (httpResponse:HttpResponse) =
    let mutable _startResponseTask : Task = null
    let _startResponseLock = obj()

    static member StartServerEventStream (httpResponse:HttpResponse, additionalHeaders:(string * string)[], cancellationToken:CancellationToken) =
        let task = backgroundTask {
            let setHeader (httpResponse:HttpResponse) (name, content:string) =
                if httpResponse.Headers.ContainsKey(name) |> not then
                    httpResponse.Headers.Add(name, StringValues(content))

            seq {
                (HeaderNames.ContentType, "text/event-stream")
                if (httpResponse.HttpContext.Request.Protocol = HttpProtocol.Http11) then
                    ("Connection", "keep-alive")
                yield! additionalHeaders
                } |> Seq.iter (setHeader httpResponse)
            do! httpResponse.StartAsync(cancellationToken)
            return! httpResponse.BodyWriter.FlushAsync(cancellationToken)
            }
        task :> Task
    static member StartServerEventStream (httpResponse, additionalHeader) = ServerSentEventHttpHandler.StartServerEventStream(httpResponse, additionalHeader, httpResponse.HttpContext.RequestAborted)

    static member SendServerEvent (httpResponse:HttpResponse, sse, cancellationToken:CancellationToken) =
        let task = task {
            let serializedSse = sse |> ServerSentEvent.serializeAsBytes |> Seq.toArray
            return! httpResponse.BodyWriter.WriteAsync(serializedSse, cancellationToken)
            }
        task :> Task
    static member SendServerEvent (sse, httpResponse) = ServerSentEventHttpHandler.SendServerEvent(httpResponse, sse, httpResponse.HttpContext.RequestAborted)

    interface ISendServerEvent with

        member this.StartServerEventStream (additionalHeaders, cancellationToken) =
            lock _startResponseLock (fun () -> if _startResponseTask = null then _startResponseTask <- ServerSentEventHttpHandler.StartServerEventStream(httpResponse, additionalHeaders, cancellationToken))
            _startResponseTask
        member this.StartServerEventStream(additionalHeaders) = (this:>ISendServerEvent).StartServerEventStream(additionalHeaders, httpResponse.HttpContext.RequestAborted)
        member this.StartServerEventStream() = (this:>ISendServerEvent).StartServerEventStream(Array.empty, httpResponse.HttpContext.RequestAborted)

        member this.SendServerEvent(sse, cancellationToken) = task {
            do! (this :> ISendServerEvent).StartServerEventStream(Array.empty, cancellationToken)
            return! ServerSentEventHttpHandler.SendServerEvent(httpResponse, sse, cancellationToken)
            }
        member this.SendServerEvent(sse) = (this:>ISendServerEvent).SendServerEvent(sse, httpResponse.HttpContext.RequestAborted)

/// Implementation of IReadSignals, for reading the Signals from the HttpRequest
[<Sealed>]
type SignalsHttpHandler (httpRequest:HttpRequest) =

    static member GetSignalsStream (httpRequest:HttpRequest) =
        match httpRequest.Method with
        | System.Net.WebRequestMethods.Http.Get ->
            match httpRequest.Query.TryGetValue(Consts.DatastarKey) with
            | true, stringValues when stringValues.Count > 0 -> (new MemoryStream(Encoding.UTF8.GetBytes(stringValues[0])) :> Stream)
            | _ -> Stream.Null
        | _ -> httpRequest.Body

    static member ReadSignalsAsync (httpRequest:HttpRequest, cancellationToken:CancellationToken) = task {
        match httpRequest.Method with
        | System.Net.WebRequestMethods.Http.Get ->
            match httpRequest.Query.TryGetValue(Consts.DatastarKey) with
            | true, stringValues when stringValues.Count > 0 -> return (stringValues[0] |> Signals.create)
            | _ -> return Signals.empty
        | _ ->
            try
                use readResult = new StreamReader(httpRequest.Body)
                let! signals = readResult.ReadToEndAsync(cancellationToken)
                return (signals |> Signals.create)
            with _ -> return Signals.empty
        }

    static member ReadSignalsAsync (httpRequest:HttpRequest) = SignalsHttpHandler.ReadSignalsAsync(httpRequest, httpRequest.HttpContext.RequestAborted)

    static member ReadSignalsAsync<'T> (httpRequest:HttpRequest, jsonSerializerOptions:JsonSerializerOptions, cancellationToken:CancellationToken) = task {
        try
            match httpRequest.Method with
            | System.Net.WebRequestMethods.Http.Get ->
                match httpRequest.Query.TryGetValue(Consts.DatastarKey) with
                | true, stringValues when stringValues.Count > 0 ->
                    return ValueSome (JsonSerializer.Deserialize<'T>(stringValues[0], jsonSerializerOptions))
                | _ ->
                    return ValueNone
            | _ ->
                let! t = JsonSerializer.DeserializeAsync<'T>(httpRequest.Body, jsonSerializerOptions, cancellationToken)
                return (ValueSome t)
        with _ -> return ValueNone
        }

    static member ReadSignalsAsync<'T> (httpRequest:HttpRequest, cancellationToken:CancellationToken) = SignalsHttpHandler.ReadSignalsAsync<'T>(httpRequest, JsonSerializerOptions.SignalsDefault, cancellationToken)
    static member ReadSignalsAsync<'T> (httpRequest:HttpRequest, jsonSerializerOptions:JsonSerializerOptions) = SignalsHttpHandler.ReadSignalsAsync<'T>(httpRequest, jsonSerializerOptions, httpRequest.HttpContext.RequestAborted)
    static member ReadSignalsAsync<'T> (httpRequest:HttpRequest) = SignalsHttpHandler.ReadSignalsAsync<'T>(httpRequest, JsonSerializerOptions.SignalsDefault)

    interface IReadSignals with
        member this.GetSignalsStream() = SignalsHttpHandler.GetSignalsStream(httpRequest)
        member this.ReadSignalsAsync(): Task<Signals> = SignalsHttpHandler.ReadSignalsAsync(httpRequest)
        member this.ReadSignalsAsync(cancellationToken: CancellationToken): Task<Signals> = SignalsHttpHandler.ReadSignalsAsync(httpRequest, cancellationToken)
        member this.ReadSignalsAsync<'T>(): Task<'T voption> = SignalsHttpHandler.ReadSignalsAsync<'T>(httpRequest)
        member this.ReadSignalsAsync<'T>(jsonSerializerOptions: JsonSerializerOptions): Task<'T voption> = SignalsHttpHandler.ReadSignalsAsync<'T>(httpRequest, jsonSerializerOptions)
        member this.ReadSignalsAsync<'T>(jsonSerializerOptions, cancellationToken) = SignalsHttpHandler.ReadSignalsAsync<'T>(httpRequest, jsonSerializerOptions, cancellationToken)